File: strata-rs/src/encode.rs
pub fn encode_uleb128(mut value: u64, out: &mut Vec<u8>) {
    loop {
        let mut byte = (value & 0x7F) as u8;
        value >>= 7;
        if value != 0 {
            byte |= 0x80;
        }
        out.push(byte);
        if value == 0 {
            break;
        }
    }
}

pub fn encode_sleb128(mut value: i64, out: &mut Vec<u8>) {
    loop {
        let byte = (value & 0x7F) as u8;
        let sign_bit = byte & 0x40;
        value >>= 7;

        let done = (value == 0 && sign_bit == 0)
            || (value == -1 && sign_bit != 0);

        out.push(if done { byte } else { byte | 0x80 });

        if done {
            break;
        }
    }
}

use crate::value::Value;

pub fn encode_value(value: &Value) -> Vec<u8> {
    let mut out = Vec::new();
    encode_into(value, &mut out);
    out
}

fn encode_into(value: &Value, out: &mut Vec<u8>) {
    match value {
        Value::Null => {
            out.push(0x00);
        }

        Value::Bool(false) => {
            out.push(0x01);
        }

        Value::Bool(true) => {
            out.push(0x02);
        }

        Value::Int(number) => {
            out.push(0x10);
            encode_sleb128(*number, out);
        }

        Value::String(string) => {
            out.push(0x20);
            let bytes = string.as_bytes();
            encode_uleb128(bytes.len() as u64, out);
            out.extend_from_slice(bytes);
        }

        Value::Bytes(bytes) => {
            out.push(0x21);
            encode_uleb128(bytes.len() as u64, out);
            out.extend_from_slice(bytes);
        }

        Value::List(items) => {
            out.push(0x30);
            encode_uleb128(items.len() as u64, out);
            for item in items {
                encode_into(item, out);
            }
        }

        Value::Map(map) => {
            out.push(0x40);
            encode_uleb128(map.len() as u64, out);

            // BTreeMap guarantees canonical order
            for (key, value) in map {
                // key encoded exactly like a String
                out.push(0x20);
                let key_bytes = key.as_bytes();
                encode_uleb128(key_bytes.len() as u64, out);
                out.extend_from_slice(key_bytes);

                encode_into(value, out);
            }
        }
    }
}

File: strata-js/src/encode.ts
import { Value } from "./value.js";
import { encodeULEB128, encodeSLEB128 } from "./varint.js";

const te = new TextEncoder();

const TAG_NULL = 0x00;
const TAG_FALSE = 0x01;
const TAG_TRUE = 0x02;

const TAG_INT = 0x10;
const TAG_STRING = 0x20;
const TAG_BYTES = 0x21;

const TAG_LIST = 0x30;
const TAG_MAP = 0x40;

export function encodeValue(value: Value): Uint8Array {
    const out: number[] = [];
    encodeInto(value, out);
    return Uint8Array.from(out);
}

function encodeInto(value: Value, out: number[]): void {
    switch (value.kind) {
        case "null": {
            out.push(TAG_NULL);
            return;
        }

        case "bool": {
            out.push(value.value ? TAG_TRUE : TAG_FALSE);
            return;
        }

        case "int": {
            out.push(TAG_INT);
            pushBytes(out, encodeSLEB128(value.value));
            return;
        }

        case "string": {
            out.push(TAG_STRING);
            const encoded = te.encode(value.value);
            pushBytes(out, encodeULEB128(BigInt(encoded.length)));
            pushBytes(out, encoded);
            return;
        }

        case "bytes": {
            out.push(TAG_BYTES);
            pushBytes(out, encodeULEB128(BigInt(value.value.length)));
            pushBytes(out, value.value);
            return;
        }

        case "list": {
            out.push(TAG_LIST);
            pushBytes(out, encodeULEB128(BigInt(value.value.length)));
            for (const item of value.value) {
                encodeInto(item, out);
            }
            return;
        }

        case "map": {
            out.push(TAG_MAP);

            // Canonical ordering
            const entries = [...value.value.entries()];
            entries.sort(([a], [b]) => compareUtf8Bytes(a, b));

            pushBytes(out, encodeULEB128(BigInt(entries.length)));

            for (const [key, value] of entries) {
                // key encoded exactly like a String
                out.push(TAG_STRING);
                const keyBytes = te.encode(key);
                pushBytes(out, encodeULEB128(BigInt(keyBytes.length)));
                pushBytes(out, keyBytes);

                encodeInto(value, out);
            }

            return;
        }
    }
}

function pushBytes(out: number[], bytes: Uint8Array): void {
    for (const b of bytes) {
        out.push(b);
    }
}

// Compare keys by UTF-8 byte order (lexicographic)
function compareUtf8Bytes(a: string, b: string): number {
    const ab = te.encode(a);
    const bb = te.encode(b);

    const min = Math.min(ab.length, bb.length);

    for (let i = 0; i < min; i++) {
        const diff = ab[i]! - bb[i]!;
        if (diff !== 0) {
            return diff;
        }
    }

    return ab.length - bb.length;
}
